File: tests/bindings/kotlin/src/test/kotlin/ConvertTest.kt
import uniffi.html2markdown_uniffi.*
import java.io.File
import kotlin.test.*
import org.junit.jupiter.api.Nested
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

class ConvertTest {

    @Nested inner class Convert {

        @Test fun `heading`() {
            assertEquals("# Hello\n", convert("<h1>Hello</h1>"))
        }

        @Test fun `empty string`() {
            assertEquals("", convert(""))
        }

        @Test fun `paragraph`() {
            assertEquals("Hello\n", convert("<p>Hello</p>"))
        }

        @Test fun `emphasis`() {
            assertEquals("*Hello World.*\n", convert("<em>Hello World.</em>"))
        }

        @Test fun `strong`() {
            assertEquals("**Hello World.**\n", convert("<strong>Hello World.</strong>"))
        }

        @Test fun `link`() {
            val html = """<a href="http://example.com" title="example">example</a>"""
            assertEquals("[example](http://example.com \"example\")\n", convert(html))
        }

        @Test fun `image`() {
            val html = """<img src="http://example.com" alt="example">"""
            assertEquals("![example](http://example.com)\n", convert(html))
        }

        @Test fun `code`() {
            assertEquals("`toString()`\n", convert("<code>toString()</code>"))
        }

        @Test fun `blockquote`() {
            val html = "<blockquote><p>This is a blockquote.</p></blockquote>"
            assertEquals("> This is a blockquote.\n", convert(html))
        }

        @Test fun `unordered list`() {
            val html = "<ul><li>Alpha</li><li>Bravo</li><li>Charlie</li></ul>"
            assertEquals("* Alpha\n* Bravo\n* Charlie\n", convert(html))
        }

        @Test fun `ordered list`() {
            val html = "<ol><li>Alpha</li><li>Bravo</li><li>Charlie</li></ol>"
            assertEquals("1. Alpha\n2. Bravo\n3. Charlie\n", convert(html))
        }
    }

    @Nested inner class ConvertWith {

        @Test fun `default options matches convert`() {
            val html = "<h1>Hello</h1>"
            val expected = convert(html)
            val actual = convertWith(html, defaultOptions())
            assertEquals(expected, actual)
        }
    }

    @Nested inner class DefaultOptions {

        @Test fun `stringify options`() {
            val opts = defaultStringifyOptions()
            assertEquals(HeadingStyle.ATX, opts.headingStyle)
            assertEquals("*", opts.bullet)
            assertEquals(".", opts.bulletOrdered)
            assertEquals("*", opts.emphasis)
            assertEquals("*", opts.strong)
            assertEquals("`", opts.fence)
            assertEquals("*", opts.rule)
            assertEquals(3.toUByte(), opts.ruleRepetition)
            assertFalse(opts.ruleSpaces)
            assertFalse(opts.closeAtx)
            assertEquals(ListItemIndent.ONE, opts.listItemIndent)
            assertTrue(opts.incrementListMarker)
            assertEquals("\"", opts.quote)
            assertTrue(opts.fences)
            assertFalse(opts.resourceLink)
        }

        @Test fun `conversion options`() {
            val opts = defaultOptions()
            assertFalse(opts.newlines)
            assertNull(opts.checked)
            assertNull(opts.unchecked)
            assertEquals(listOf("\""), opts.quotes)
        }
    }

    @Nested inner class ErrorHandling {

        @Test fun `invalid bullet throws OptionsException`() {
            val opts = defaultOptions().copy(
                stringify = defaultStringifyOptions().copy(bullet = "x")
            )
            val ex = assertFailsWith<OptionsException.InvalidOption> {
                convertWith("<p>hi</p>", opts)
            }
            assertEquals("bullet", ex.field)
            assertEquals("x", ex.value)
        }

        @Test fun `empty bullet throws OptionsException`() {
            val opts = defaultOptions().copy(
                stringify = defaultStringifyOptions().copy(bullet = "")
            )
            assertFailsWith<OptionsException.InvalidOption> {
                convertWith("<p>hi</p>", opts)
            }
        }
    }

    @Nested inner class Fixtures {

        private val fixturesDir: File by lazy {
            val path = System.getProperty("fixtures.dir")
                ?: error("System property 'fixtures.dir' not set — run via Gradle")
            File(path).also {
                assertTrue(it.isDirectory, "Fixtures directory not found: $it")
            }
        }

        @ParameterizedTest
        @ValueSource(strings = [
            "a", "blockquote", "br", "code", "em", "heading",
            "img", "ol", "paragraph", "strong", "table", "ul",
        ])
        fun `matches expected markdown`(name: String) {
            val dir = fixturesDir.resolve(name)
            assertTrue(dir.isDirectory, "Fixture directory not found: $dir")

            val html = dir.resolve("index.html").readText()
            val expectedMd = dir.resolve("index.md").readText()

            val result = convert(html)
            assertEquals(expectedMd, result, "Fixture '$name' mismatch")
        }
    }
}

File: tests/bindings/dart/test/convert_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:html2markdown_dart_test/html2markdown_uniffi.dart';

void main() {
  setUpAll(() {
    // Load the native library from the build output.
    final projectRoot = Directory.current.path.replaceAll(
      RegExp(r'/tests/bindings/dart$'),
      '',
    );
    final libDir = '$projectRoot/uniffi/target/release';
    final String libPath;
    if (Platform.isMacOS) {
      libPath = '$libDir/libhtml2markdown_uniffi.dylib';
    } else if (Platform.isLinux) {
      libPath = '$libDir/libhtml2markdown_uniffi.so';
    } else {
      throw UnsupportedError('Unsupported platform: ${Platform.operatingSystem}');
    }
    configureDefaultBindings(libraryPath: libPath);
  });

  group('convert', () {
    test('heading', () {
      expect(convert('<h1>Hello</h1>'), equals('# Hello\n'));
    });

    test('empty string', () {
      expect(convert(''), equals(''));
    });

    test('paragraph', () {
      expect(convert('<p>Hello</p>'), equals('Hello\n'));
    });

    test('emphasis', () {
      expect(convert('<em>Hello World.</em>'), equals('*Hello World.*\n'));
    });

    test('strong', () {
      expect(convert('<strong>Hello World.</strong>'), equals('**Hello World.**\n'));
    });

    test('link', () {
      final html = '<a href="http://example.com" title="example">example</a>';
      expect(convert(html), equals('[example](http://example.com "example")\n'));
    });

    test('image', () {
      final html = '<img src="http://example.com" alt="example">';
      expect(convert(html), equals('![example](http://example.com)\n'));
    });

    test('code', () {
      expect(convert('<code>toString()</code>'), equals('`toString()`\n'));
    });

    test('blockquote', () {
      final html = '<blockquote><p>This is a blockquote.</p></blockquote>';
      expect(convert(html), equals('> This is a blockquote.\n'));
    });

    test('unordered list', () {
      final html = '<ul><li>Alpha</li><li>Bravo</li><li>Charlie</li></ul>';
      expect(convert(html), equals('* Alpha\n* Bravo\n* Charlie\n'));
    });

    test('ordered list', () {
      final html = '<ol><li>Alpha</li><li>Bravo</li><li>Charlie</li></ol>';
      expect(convert(html), equals('1. Alpha\n2. Bravo\n3. Charlie\n'));
    });
  });

  group('convertWith', () {
    test('default options matches convert', () {
      final html = '<h1>Hello</h1>';
      expect(convertWith(html, defaultOptions()), equals(convert(html)));
    });
  });

  group('defaultOptions', () {
    test('stringify options', () {
      final opts = defaultStringifyOptions();
      expect(opts.headingStyle, equals(HeadingStyle.atx));
      expect(opts.bullet, equals('*'));
      expect(opts.bulletOrdered, equals('.'));
      expect(opts.emphasis, equals('*'));
      expect(opts.strong, equals('*'));
      expect(opts.fence, equals('`'));
      expect(opts.rule, equals('*'));
      expect(opts.ruleRepetition, equals(3));
      expect(opts.ruleSpaces, isFalse);
      expect(opts.closeAtx, isFalse);
      expect(opts.listItemIndent, equals(ListItemIndent.one));
      expect(opts.incrementListMarker, isTrue);
      expect(opts.quote, equals('"'));
      expect(opts.fences, isTrue);
      expect(opts.resourceLink, isFalse);
    });

    test('conversion options', () {
      final opts = defaultOptions();
      expect(opts.newlines, isFalse);
      expect(opts.checked, isNull);
      expect(opts.unchecked, isNull);
      expect(opts.quotes, equals(['"']));
    });
  });

  group('error handling', () {
    test('invalid bullet throws OptionsErrorException', () {
      final opts = defaultOptions().copyWith(
        stringify: defaultStringifyOptions().copyWith(bullet: 'x'),
      );
      expect(
        () => convertWith('<p>hi</p>', opts),
        throwsA(
          isA<OptionsErrorExceptionInvalidOption>()
            .having((e) => e.field, 'field', equals('bullet'))
            .having((e) => e.value, 'value', equals('x')),
        ),
      );
    });

    test('empty bullet throws OptionsErrorException', () {
      final opts = defaultOptions().copyWith(
        stringify: defaultStringifyOptions().copyWith(bullet: ''),
      );
      expect(
        () => convertWith('<p>hi</p>', opts),
        throwsA(isA<OptionsErrorExceptionInvalidOption>()),
      );
    });
  });

  group('fixtures', () {
    final projectRoot = Directory.current.path.replaceAll(
      RegExp(r'/tests/bindings/dart$'),
      '',
    );
    final fixturesDir = Directory('$projectRoot/test-fixtures');

    const fixtureNames = [
      'a', 'blockquote', 'br', 'code', 'em', 'heading',
      'img', 'ol', 'paragraph', 'strong', 'table', 'ul',
    ];

    for (final name in fixtureNames) {
      test('fixture: $name', () {
        final dir = Directory('${fixturesDir.path}/$name');
        expect(dir.existsSync(), isTrue, reason: 'Fixture dir not found: $dir');

        final html = File('${dir.path}/index.html').readAsStringSync();
        final expectedMd = File('${dir.path}/index.md').readAsStringSync();
        final configJson = jsonDecode(
          File('${dir.path}/index.json').readAsStringSync(),
        ) as Map<String, dynamic>;

        if (configJson['fragment'] != true) {
          // Skip non-fragment fixtures
          return;
        }

        expect(convert(html), equals(expectedMd),
            reason: "Fixture '$name' mismatch");
      });
    }
  });
}
